### PR TITLE
proposed option fetch from Tradier that is aware of the latest minute intraday pricing so that the in-the-money, at-the-money and out-of-the-money options are limited to [current close - 10 strikes, atm the money strike, current close + 10 strikes]. This helps to reduce the amount of data being stored in redis for intraday option chains.

### PR DESCRIPTION
**Pull request by jay-johnson**

### CODE REVIEW

File: analysis_engine/td/extract_df_from_redis.py [hunk 1]
```diff
@@ -44,46 +44,193 @@ def extract_option_calls_dataset(
...
+    if verbose:
+        log.info(
+            f'{label} - {df_str} - start - redis_key={redis_key} '
+            f's3_key={s3_key}')
 
-def extract_option_puts_dataset(
-        work_dict,
-        scrub_mode='sort-by-date'):
-    """extract_option_puts_dataset
+    if not redis_host and not redis_port:
```
Comment (virdesai) [inline analysis_engine/td/extract_df_from_redis.py]: should we separate these checks?

File: analysis_engine/td/extract_df_from_redis.py [hunk 2]
```diff
@@ -44,46 +44,193 @@ def extract_option_calls_dataset(
...
+                calls_df = pd.read_json(
+                    calls_json,
+                    orient='records')
+                if len(calls_df.index) == 0:
+                    return ae_consts.SUCCESS, None
+                if 'date' not in calls_df:
+                    log.debug(
+                        'failed to find date column in TD calls '
+                        f'df={calls_df} from lens={len(calls_df.index)}')
+                    return ae_consts.SUCCESS, None
```
Comment (virdesai) [inline analysis_engine/td/extract_df_from_redis.py]: success if no date column?

File: analysis_engine/td/extract_df_from_redis.py [hunk 3]
```diff
@@ -44,46 +44,193 @@ def extract_option_calls_dataset(
...
+                exp_date_str = (
+                    calls_df['exp_date'].iloc[-1])
+
+                calls_df['date'] = calls_df['date'].dt.strftime(
+                    ae_consts.COMMON_TICK_DATE_FORMAT)
+
+            except Exception as f:
+                not_fixed = True
+                if (
+                        f'Can only use .dt accessor with '
```
Comment (virdesai) [inline analysis_engine/td/extract_df_from_redis.py]: don't need to use f-string if no variable/function is being inserted

File: analysis_engine/td/extract_df_from_redis.py [hunk 4]
```diff
@@ -44,46 +44,193 @@ def extract_option_calls_dataset(
...
+                        f'datetimelike values') in str(f):
```
Comment (virdesai) [inline analysis_engine/td/extract_df_from_redis.py]: `datetimelike` or `datetime like`?

File: analysis_engine/td/extract_df_from_redis.py [hunk 5]
```diff
@@ -44,46 +44,193 @@ def extract_option_calls_dataset(
...
+                    try:
+                        log.critical(
+                            f'fixing dates in {redis_key}')
+                        # remove epoch second data and
+                        # use only the millisecond date values
+                        calls_df['date'][calls_df['date'] < 1600000] = None
```
Comment (virdesai) [inline analysis_engine/td/extract_df_from_redis.py]: just as a general note that Ara beat into me, 1600000 seems like a magic number. what's it supposed to mean? is it a const used in multiple files that can be declared in 1 place or a local const that can be defined at the top outside of the functions?

**Comment by virdesai** [comment]
added some comments but basically looks good. I can refactor the functions back together if I want but wasn't gaining us anything other than less total lines of code so it's not a big deal either way

**Review by virdesai — APPROVED** [review]

File: analysis_engine/td/fetch_api.py [hunk 6]
```diff
@@ -30,49 +30,187 @@ def fetch_calls(
...
+    res = url_helper.url_helper(sess=session).get(
+        use_url
+    )
 
-    Fetch the Tradier puts intraday data for a ticker and
-    return it as a ``pandas.DataFrame``.
+    if res.status_code != requests.codes.OK:
+        if res.status_code in [401, 403]:
+            log.critical(
+                f'Please check the TD_TOKEN is correct '
```
Comment (virdesai) [inline analysis_engine/td/fetch_api.py]: don't need to use f-string but doesn't really hurt

File: analysis_engine/td/fetch_api.py [hunk 7]
```diff
@@ -30,49 +30,187 @@ def fetch_calls(
...
+        if (
+                node['option_type'] == 'call' and
+                node['expiration_type'] == 'standard' and
+                float(node['bid']) > 0.01):
+            node['opt_type'] = int(ae_consts.OPTION_CALL)
+            node['exp_date'] = node['expiration_date']
+
+            new_node = {}
+            for col in td_consts.TD_OPTION_COLUMNS:
+                if col in node:
```
Comment (virdesai) [inline analysis_engine/td/fetch_api.py:144]: is there a way to do a `for col in td_consts.TD_OPTION_COLUMNS and in node`? just wondering if it's possible, doesn't save anything for us or help anything here

File: analysis_engine/td/extract_df_from_redis.py (hunk 1 above)
Comment (virdesai) [reply analysis_engine/td/extract_df_from_redis.py]: actually could we just set these as the defaults of the work_dict get method?

File: analysis_engine/td/extract_df_from_redis.py [hunk 8]
```diff
@@ -44,46 +44,193 @@ def extract_option_calls_dataset(
...
+                f'{label} - {df_str} redis get data key={redis_key} '
+                f'status={ae_consts.get_status(status=status)}')
 
+        if status == ae_consts.SUCCESS:
+            calls_json = None
+            if 'calls' in redis_rec['rec']['data']:
+                calls_json = redis_rec['rec']['data']['calls']
+            else:
+                calls_json = redis_rec['rec']['data']
+            if verbose:
```
Comment (virdesai) [inline analysis_engine/td/extract_df_from_redis.py:124]: this verbosity doesn't seem to be providing any extra info from the previously added verbose log

File: analysis_engine/td/extract_df_from_redis.py [hunk 9]
```diff
@@ -44,46 +44,193 @@ def extract_option_calls_dataset(
...
+                log.info(f'{label} - {df_str} redis convert calls to df')
+            exp_date_str = None
+            try:
+                calls_df = pd.read_json(
+                    calls_json,
+                    orient='records')
+                if len(calls_df.index) == 0:
```
Comment (virdesai) [inline analysis_engine/td/extract_df_from_redis.py]: wouldn't the pandas `read_json` method with `orient='records'` return a list? so the `.index` would be pointing to a function object and taking the len of that would be invalid? http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_json.html

File: analysis_engine/td/extract_df_from_redis.py (hunk 5 above)
Comment (virdesai) [reply analysis_engine/td/extract_df_from_redis.py]: also does this work? reading this it seems like you're setting a boolean true or false key for `calls_df['date']` object and putting a value of None on it? Is it resulting in a 0 or 1 based on true or false evaluation and that's the index that's being set?

File: analysis_engine/td/fetch_api.py [hunk 10]
```diff
@@ -102,12 +246,12 @@ def fetch(
     if res.status_code != requests.codes.OK:
         if res.status_code in [401, 403]:
             log.critical(
-                'Please check the TD_TOKEN is correct '
+                f'Please check the TD_TOKEN is correct '
```
Comment (virdesai) [inline analysis_engine/td/fetch_api.py]: don't need to use f-string but doesn't really hurt

File: analysis_engine/td/fetch_api.py [hunk 11]
```diff
@@ -102,12 +246,12 @@ def fetch(
...
                 f'received {res.status_code} during '
-                f'fetch for: {fetch_type}s')
+                f'fetch for: puts')
```
Comment (virdesai) [inline analysis_engine/td/fetch_api.py]: don't need to use f-string but doesn't really hurt

File: analysis_engine/td/fetch_api.py [hunk 12]
```diff
@@ -30,49 +30,187 @@ def fetch_calls(
...
+        ],
+        ascending=True)
+
+    num_chains = len(full_df.index)
+
+    df = None
+    if latest_close:
+        df_filter = (
+            (full_df['strike'] >=
+                (latest_close - 10.0)) &
```
Comment (virdesai) [inline analysis_engine/td/fetch_api.py]: should these `10.0`s be adjustable/a const in td.consts?

File: analysis_engine/td/fetch_api.py [hunk 13]
```diff
@@ -30,49 +30,187 @@ def fetch_calls(
...
+                (latest_close + 10.0)))
+        df = full_df[df_filter].copy().sort_values(
+            by=[
+                'date',
+                'strike'
+            ]).reset_index()
+    else:
+        mid_chain_idx = int(num_chains / 2)
+        low_idx = int(
+            mid_chain_idx - 200)
```
Comment (virdesai) [inline analysis_engine/td/fetch_api.py]: do we want this 200 creating a 400 index range to be adjustable and/or a const?

File: analysis_engine/td/extract_df_from_redis.py (hunk 4 above)
Comment (jay-johnson) [reply analysis_engine/td/extract_df_from_redis.py]: this is a pandas error and only happens for data collected prior to the migration done back in jan. if you hit the data must be corrected by backing up to aws, deleting the data in redis, then restoring from the backup. the restore was fixed a while back but this was left so i can remember how it worked if others hit it

File: analysis_engine/td/extract_df_from_redis.py (hunk 5 above)
Comment (jay-johnson) [reply analysis_engine/td/extract_df_from_redis.py]: done consts.py will have it

File: analysis_engine/td/fetch_api.py (hunk 7 above)
Comment (jay-johnson) [reply analysis_engine/td/fetch_api.py:144]: who knows. i'm sure a future pr could fix it

File: analysis_engine/td/extract_df_from_redis.py (hunk 8 above)
Comment (jay-johnson) [reply analysis_engine/td/extract_df_from_redis.py:124]: it's added due to the serialization changes made to the td options back in jan. this prob could be removed in a future version, but i used it to see the dictionaries of v1 vs the new version did not match up with the calls nested under the data dict

File: analysis_engine/td/fetch_api.py (hunk 12 above)
Comment (jay-johnson) [reply analysis_engine/td/fetch_api.py]: done

File: analysis_engine/td/fetch_api.py (hunk 13 above)
Comment (jay-johnson) [reply analysis_engine/td/fetch_api.py]: done - i think some users may have the capacity to store all option chains and this is using an env var in consts.py

**Review by jay-johnson — COMMENTED** [review]
pushing changes

File: analysis_engine/td/extract_df_from_redis.py (hunk 2 above)
Comment (jay-johnson) [reply analysis_engine/td/extract_df_from_redis.py]: using EMPTY now

**Pull request merged**